portal-bridge/src/bridge/state.rs

```diff
@@ -253,73 +219,174 @@ impl StateBridge {
         address_hash: B256,
         block_hash: B256,
     ) -> anyhow::Result<()> {
-        let permit = self
-            .gossip_limit_semaphore
-            .clone()
-            .acquire_owned()
-            .await
-            .expect("to be able to acquire semaphore");
-        let storage_content_key = create_storage_content_key(storage_proof, address_hash)?;
-        let storage_content_value =
-            create_storage_content_value(block_hash, account_proof, storage_proof)?;
-        Self::spawn_serve_state_proof(
-            self.portal_client.clone(),
-            storage_content_key,
-            storage_content_value,
-            permit,
-            self.metrics.clone(),
-        );
+        let content_key = create_storage_content_key(storage_proof, address_hash)?;
+        let content_value = create_storage_content_value(block_hash, account_proof, storage_proof)?;
+        self.spawn_offer_tasks(content_key, content_value).await;
         Ok(())
     }
 
-    fn spawn_serve_state_proof(
-        portal_client: HttpClient,
+    // request enrs interested in the content key from Census
+    async fn request_enrs(&self, content_key: &StateContentKey) -> anyhow::Result<Vec<Enr>> {
+        let (resp_tx, resp_rx) = futures::channel::oneshot::channel();
+        let enrs_request = EnrsRequest {
+            content_key: ContentKey::State(content_key.clone()),
+            resp_tx,
+        };
+        self.census_tx.send(enrs_request)?;
+        Ok(resp_rx.await?)
+    }
+
+    // spawn individual offer tasks of the content key for each interested enr found in Census
+    async fn spawn_offer_tasks(
+        &self,
         content_key: StateContentKey,
         content_value: StateContentValue,
-        permit: OwnedSemaphorePermit,
-        metrics: BridgeMetricsReporter,
-    ) -> JoinHandle<()> {
-        info!("Spawning serve_state_proof for content key: {content_key}");
-        tokio::spawn(async move {
-            let timer = metrics.start_process_timer("spawn_serve_state_proof");
-            match timeout(
-                SERVE_BLOCK_TIMEOUT,
-                Self::serve_state_proof(
-                    portal_client,
-                    content_key.clone(),
-                    content_value,
-                    metrics.clone()
-                )).await
+    ) {
+        let Ok(enrs) = self.request_enrs(&content_key).await else {
+            error!("Failed to request enrs for content key, skipping offer: {content_key:?}");
+            return;
+        };
+        self.stats
```

Comment:
Do we need `stats` field?

How about we add content key to the `OfferReport`, create new one here (`Arc<Mutex<OfferReport>>`), and pass it to tasks below (the last one that updates it will print the report)?

Up to you.